### PR TITLE
Fixes + Tweaks

### LINKED ISSUE #395: "Allow "Tag to File Name" to create new folders" does work but it will not work when "Limit File Name Characters" setting turned on.
BTW it does work very well. Using custom format string `%albumartist%/%album%/ %artist% - %title%` will give this result.

however when "Limit File Name Characters" setting turned on it will give this result.

### PR DESCRIPTION
**Pull request by nlogozzo**
Fixes #395 

TODO:

- [x] Fix Tagger ignoring space with invalid chars

WinUI Tweaks:
- [x] Removed Apply button for SyncLyricRow. Lyric will be automatically updated as the text box changes
- [x] Make add sync lyric a flyout instead of showing new ContentDialog
- [x] See where else flyouts can be used instead of full ContentDialogs

### CODE REVIEW

**Comment by kissthermite** [comment]
> @kissthermite Could you test this build: https://github.com/NickvisionApps/Tagger/actions/runs/6776895782

sure! will let you know in a moments

**Comment by kissthermite** [comment]
Using custom format string `/%albumartist%/%album%/%artist% - %title%` with `Limit File Name Characters` setting turn on does work as expected for others limited characters except `/`

---

tag to file name conversion with aforementioned format string with limited character being `:`


will give this output: `/Kotono Nagase (CV_Mirai Tachibana)/song for you(Kotono version)/Kotono Nagase (CV_Mirai Tachibana) - song for you(Kotono version).flac`


---

tag to file name conversion with aforementioned format string with limited character being `/`


will give this output: `/SQUARE ENIX MUSIC/NieR Gestalt & NieR Replicant Original Soundtrack/SQUARE ENIX MUSIC - Song of the Ancients /Devola.flac`


---

**Comment by nlogozzo** [comment]
> tag to file name conversion with aforementioned format string with limited character being /

Well now with the new algorithm `/` means new directory. But you want it where if the `/` was contained in a tag property and now the format string, to not count as a new directory, correct?

I think I can do that.

**Comment by kissthermite** [comment]
>But you want it where if the `/` was contained in a tag property and now the format string, to not count as a new directory, correct?

Correct indeed.

**Comment by nlogozzo** [comment]
@fsobolev If you could make sure the algorithm in this commit still holds: https://github.com/NickvisionApps/Tagger/pull/396/commits/325b0f408260cd0d5a69dfc756fda2e213346a14

**Comment by nlogozzo** [comment]
> > But you want it where if the `/` was contained in a tag property and now the format string, to not count as a new directory, correct?
> 
> Correct indeed.

@kissthermite Try this build: https://github.com/NickvisionApps/Tagger/actions/runs/6779510126?pr=396

P.S. Thanks for testing and sorry for messing up your library structure with these tests 🫠

**Comment by nlogozzo** [comment]
> > > But you want it where if the `/` was contained in a tag property and now the format string, to not count as a new directory, correct?
> > 
> > 
> > Correct indeed.
> 
> @kissthermite Try this build: https://github.com/NickvisionApps/Tagger/actions/runs/6779510126?pr=396
> 
> P.S. Thanks for testing and sorry for messing up your library structure with these tests 🫠

Sorry I lied, please try this build: https://github.com/NickvisionApps/Tagger/actions/runs/6779656223/job/18427058390

If a tag property contains `/`, regardless of using Limit Filename Characters, it will not create a new directory. Only if the format string explicitly contains a `/`, then it will.

**Comment by kissthermite** [comment]
> P.S. Thanks for testing and sorry for messing up your library structure with these tests 🫠

Don't worries my actual music library are stored on /var/home/kissthermite/Jellyfin Server Media/Music

The one on Downloads folder are just a copy for testing purpose XD.

**Comment by kissthermite** [comment]
> If a tag property contains `/`, regardless of using Limit Filename Characters, it will not create a new directory. Only if the format string explicitly contains a `/`, then it will.

Just tested and indeed it does behave the way you said whether `File Name Characters` turn on/off. Good job @nlogozzo! and enjoy your well deserved rest tonight.

**Comment by nlogozzo** [comment]
> Just tested and indeed it does behave the way you said whether `Limit File Name Characters` turn on/off. Good job @nlogozzo! and enjoy your well deserved rest tonight.

Yay, glad it's working as expected now! Once @fsobolev approves of the rest of this PR, I will release a beta. 

If you have any other issues/feature requests please try to report them before Wednesday that way I can implement them all on Wednesday and then get a stable out by Friday 😅

**Comment by kissthermite** [comment]
> If you have any other issues/feature requests please try to report them before Wednesday that way I can implement them all on Wednesday and then get a stable out by Friday 😅

Actually there's one thing.

With `Limit File Name Characters` turn on, Tagger ignored space before, between & after that limited characters, is that the intended behavior?

**Comment by nlogozzo** [comment]
> With Limit File Name Characters turn on, Tagger ignored space before, between & after that limited characters, is that the intended behavior?

Hmm...no it's not intended. Will take a look at it tomorrow to see what's going on.

File: NickvisionTagger.Shared/Models/MusicFile.cs [hunk 1]
```diff
@@ -886,6 +887,7 @@ public bool TagToFilename(string formatString)
                 {
                     replace = PublishingDate == DateTime.MinValue ? "" : PublishingDate.ToShortDateString();
                 }
+                replace = replace.Replace(RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? '\\' : '/', '_');
```
Comment (DaPigGuy) [inline NickvisionTagger.Shared/Models/MusicFile.cs:890]: `Path.DirectorySeparatorChar`

File: NickvisionTagger.Shared/Models/MusicFile.cs [hunk 2]
```diff
@@ -85,9 +85,10 @@ static MusicFile()
         {
             _invalidSystemFilenameCharacters.Remove('\\');
         }
-        else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+        else
         {
             _invalidSystemFilenameCharacters.Remove('/');
+            _invalidWindowsFilenameCharacters.Remove('/');
```
Comment (DaPigGuy) [inline NickvisionTagger.Shared/Models/MusicFile.cs:91]: Well, `_invalidWindowsFilenameCharacters` is only used on non-Windows platforms for the Limit Filename Characters preference anyway, so maybe just not include `/` :sweat_smile:

Comment (DaPigGuy) [reply NickvisionTagger.Shared/Models/MusicFile.cs:91]: `_invalidSystemFilenameCharacters.Remove(Path.DirectorySeparatorChar);`

**Review by DaPigGuy — CHANGES_REQUESTED** [review]

**Comment by nlogozzo** [comment]
>With Limit File Name Characters turn on, Tagger ignored space before, between & after that limited characters, is that the intended behavior?

@kissthermite Can't reproduce...
Here's WinUI:
image
Here's GNOME:
image

I think it just seems every small in GNOME but the space is indeed there.

**Comment by kissthermite** [comment]
Yeah. My bad just check it again and indeed it's just a very tiny space.

**Comment by nlogozzo** [comment]
I was going to release a beta but decided against it since it's almost Friday and if any bugs come up before then I will fix them and just release stable. Since your issue was confirmed to be fixed, no need for a beta.

**Pull request merged**